---
title: "Ligand and receptor contribution to change in communication profile"
---

In this notebook we explore the ligand and receptor contributions to the change in the communication profile. We are interested in:

 - how much do receptors and ligands contribute to the change? (e.g. for the topmost interactions identified by ORA)
 - is the pattern global or does the contribution depend on context? (e.g. particular tissue and celltypes)
 - do we observe saturation behavior and signs of down-regulation? As the concentration of ligand increases, many celltypes will down-regulate the expression of their receptor
 - is there a bias in LRs with multiligand or multireceptor components?

# Getting the prerequisite data and results

We'll need the interactions dataframe and ora results to prioritize the interactions we're examining during the exploration.

## 1) Get data

```{r}
library(data.table)
library(purrr)
library(ggplot2)


RESULTS_PATH = '../../data_scAgeCom/testing/scdiffcom_tms_facs_mixed_Spleen.rds'
results = readRDS(RESULTS_PATH)

source('../src/utils.R')
results = add_dummy_tissue(results)

str(results, max.level = 1)

```

## 2) Filtering and ORA:

Top ORA LR interactions:
```{r}
top_ora = results$ORA[
    Category == 'LR_GENES' 
    & pval_adjusted_DIFF < 0.05 
    & Tissue == 'DummyTissue'][
        order(-OR), .(
            Value, 
            OR_DIFF, pval_adjusted_DIFF,
            OR_UP, pval_adjusted_UP,
            OR_DOWN, pval_adjusted_DOWN)
        ]
top_lr = top_ora[, Value][1:30]
top_lr_up = top_ora[order(-OR_UP), Value][1:30]
top_ora
```


# EDA on selected ligand receptor

## Add ligand and receptor-change measurement

First let's add difference in ligand and difference in receptor activity with age. We have log-transformed average expression value for ligand and receptor gene for young and old. Therefore, to obtain logFC between gene expression we take the difference (e.g. L_EXPRESSION_old - L_EXPRESSION_young). A **caveat** though is that the log-transformation includes pseudocounts, i.e. L_EXPRESSION_old = average(log(expression_i + 1)). We'll ignore for the time being.

```{r}
# results$scdiffcom_dt_filtered[
#     , 
#     c('L_DIFF', 'R_DIFF') := .(
#         max(
#             L1_EXPRESSION_OLD - L1_EXPRESSION_YOUNG,
#             L2_EXPRESSION_OLD - L2_EXPRESSION_YOUNG
#         ),
#         
#         max(
#             R1_EXPRESSION_OLD - R1_EXPRESSION_YOUNG,
#             R2_EXPRESSION_OLD - R2_EXPRESSION_YOUNG,
#             R3_EXPRESSION_OLD - R3_EXPRESSION_YOUNG
#         )
#     )
# ]

results$scdiffcom_dt_filtered[
    ,
    c('L_DIFF', 'R_DIFF') := .(
        L1_EXPRESSION_OLD - L1_EXPRESSION_YOUNG,
        R1_EXPRESSION_OLD - R1_EXPRESSION_YOUNG
        )
]


```

## Explore graphically ligand and receptor contributions

We'll look at LR contributions across tissues and celltypes for top interactions.

Let's look at distribution of L and R change. We notice that the distributions are right-skewed.

```{r}
dt = results$scdiffcom_dt_filtered

plot_histo <- function(dt, col, title) {
    (ggplot(data=dt, aes(get(col)))
    + geom_histogram(aes(color=DIFFERENTIALLY_EXPRESSED))
    + ggtitle(title))
}

plot_histo(dt, "L_DIFF", "Ligand change distribution")
plot_histo(dt, "R_DIFF", "Receptor change distribution")
plot_histo(dt, "LOGFC", "Ligand*Receptor change distribution")
```

Let's look at the global and per tissue pattern of LR change. In the right plot corresponding to interactions that are differentially expressed with age, the lower-left quadrant appears to have more mass than the upper-right, concordant with the above observation that there are more interactions going down with age rather than up.

```{r}

plot_global_density <- function(dt) {
    global_density = (
        ggplot(data=dt, aes(L_DIFF, R_DIFF))
        + geom_point(size=1, alpha=0.5)
        + geom_density2d_filled(alpha=0.7, bins=20)
        + geom_hline(yintercept=0, linetype='dashed', color='red', size=0.25)
        + geom_vline(xintercept=0, linetype='dashed', color='red', size=0.25)
        + ggtitle("Global L and R change profile")
        + theme_minimal()
        + facet_grid(. ~ DIFFERENTIALLY_EXPRESSED)
        + theme(legend.position="none")
    )
    return(global_density)
}

plot_global_density(dt)


# byTissue_density = (
#     ggplot(data=dt, aes(L_DIFF, R_DIFF))
#     # + geom_point(size=0.05, alpha=0.6)
#     + geom_bin2d(binwidth=0.1)
#     + geom_hline(yintercept=0, linetype='dashed', color='red', size=0.2)
#     + geom_vline(xintercept=0, linetype='dashed', color='red', size=0.2)
#     + ggtitle("By tissue L and R change profile")
#     + theme_minimal()
#     + facet_grid(TISSUE ~ DIFF_EXPRESSION)
#     + coord_fixed()
# )
# ggsave("testplot.png", byTissue_density, width=30, height=60, units='cm')
```


Let's check same celltype communication change


```{r}
dt_auto = dt[L_CELLTYPE == R_CELLTYPE]

plot_histo(dt_auto, "L_DIFF", "Ligand change distribution")
plot_histo(dt_auto, "R_DIFF", "Receptor change distribution")
plot_histo(dt_auto, "LOGFC", "Ligand*Receptor change distribution")
plot_global_density(dt_auto)
```


Let's look into some particular interactions. I would note that changes in interactions predominantly due to the ligand-only or receptor-only are detectable by studies that look at DGE. It may be more robust and valuable to look at LR interactions where both the ligand and the receptor get's down.


```{r}
# plot_interactions <- function(dt, selected_interactions, filename='test.png') {
#     plot_interacts = (
#         ggplot(data=dt[LR_GENES %in% selected_interactions 
#                        & DIFF_EXPRESSION == TRUE],
#                aes(L_DIFF, R_DIFF))
#         + geom_point(aes(color=TISSUE), size=1.5)  # shape=DIFF_EXPRESSION
#         + coord_fixed()
#         + geom_hline(yintercept=0, linetype='dashed', color='red', size=0.25)
#         + geom_vline(xintercept=0, linetype='dashed', color='red', size=0.25)
#         + ggtitle("Selected interactions")
#         + theme_minimal()
#         + facet_wrap(. ~ LR_GENES)
#     )
#     ggsave(filename, plot_interacts, width=50, height=50, units='cm')   
# }
# plot_interactions(dt, top_lr, 'top_lr.png')
# plot_interactions(dt, top_lr_up, 'top_lr_up.png')
```


# Systematic approach